designs/extensible-cli-and-scaffolding-plugins-phase-1-5.md

```diff
@@ -0,0 +1,258 @@
+This method will be used to inject the `Config` object that the plugin can modify at will.
+The CLI will create/load/save this configuration object.
+
+- Required/optional
+  - [ ] Required
+  - [x] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+
+#### Inject resource
+This method will be used to inject the `Resource` object.
+
+- Required/optional
+  - [x] Required
+  - [ ] Optional
+- Subcommands
+  - [ ] Init
+  - [ ] Edit
+  - [x] Create API
+  - [x] Create webhook
+
+#### Pre-scaffold
+This method will be used to take actions before the main scaffolding is performed, e.g. validations.
+
+NOTE: a filesystem abstraction will be passed to this method that must be used for scaffolding.
+
+- Required/optional
+  - [ ] Required
+  - [x] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+
+#### Scaffold
+This method will be used to perform the main scaffolding.
+
+NOTE: a filesystem abstraction will be passed to this method that must be used for scaffolding.
+
+- Required/optional
+  - [x] Required
+  - [ ] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+
+#### Post-scaffold
+This method will be used to take actions after the main scaffolding is performed, e.g. cleanup.
+
+NOTE: a filesystem abstraction will **NOT** be passed to this method, as post-scaffold task do not require it.
+In case some post-scaffold task requires a filesystem abstraction, it could be added.
+
+- Required/optional
+  - [ ] Required
+  - [x] Optional
+- Subcommands
+  - [x] Init
+  - [x] Edit
+  - [x] Create API
+  - [x] Create webhook
+
+## Implementation
+
+The following types are used as input/output values of the described methods:
+```go
+// CLIMetadata is the runtime meta-data of the CLI
+type CLIMetadata struct {
+	// CommandName is the root command name.
+	CommandName string
+}
+
+// SubcommandMetadata is the runtime meta-data for a subcommand
+type SubcommandMetadata struct {
+	// Description is a description of what this subcommand does. It is used to display help.
+	Description string
+	// Examples are one or more examples of the command-line usage of this subcommand. It is used to display help.
+	Examples string
+}
+
+type ExitError struct {
+	Plugin string
+	Reason string
+}
+
+func (e ExitError) Error() string {
+	return fmt.Sprintf("plugin %s exit early: %s", e.Plugin, e.Reason)
+}
+```
+
+The described methods are implemented through the use of the following interfaces.
+```go
+type RequiresCLIMetadata interface {
+	InjectCLIMetadata(CLIMetadata)
+}
+
+type UpdatesSubcommandMetadata interface {
+	UpdateSubcommandMetadata(*SubcommandMetadata)
+}
+
+type HasFlags interface {
+	BindFlags(*pflag.FlagSet)
+}
+
+type RequiresConfig interface {
+	InjectConfig(config.Config) error
+}
+
+type RequiresResource interface {
+	InjectResource(*resource.Resource) error
+}
+
+type HasPreScaffold interface {
```

Comment:
I don't see an immediate need for pre and post scaffolding, since these can be done within `Scaffold()`. Perhaps remove these to simplify things, and add them back if we find a valid use case for them?

Reply:
They are optional, so there is no need to use them if you want to develop a simple plugin (you can check the edit subcommands in the implementation PR), but they do offer a great flexibility. All plugins will run their pre-scaffold logic and can return an error (stoping the subcomman execution) before any of the plugins start to scaffold the files. This allows us to validate all the plugins before scaffolding. If we only had one hook (Scaffold) then every plugin will run its validation, scaffold its files and do the post-scaffolding tasks before the rest of the plugin can validate theirselves. Same applies for post-scaffolding tasks.

Let me illustrate this with a couple examples:
1. We have a plugin that allows us to define the fields a type will have in some yaml format that must be present before we run `kb create api`. The cli has the base go/v3 and this plugins configured for this project. One of the validations of this plugins is to check that the file is present and that it can be parsed. If we only have one hook, it will check this conditions after go/v3 has scaffolded its files, when it shouldn't because it failed to fulfill the second plugin requirements.
2. We have a plugin that adds some fields to the generated types. Again, the cli has the base go/v3 plugin and this one that edits its fields. go/v3 runs `make generate` at the end, but if we don't have a post-scaffold hook, it will be run before the second plugin performs the changes to the types file, and therefore be obsolete by the time the command finishes.

Reply:
Thanks for the explanation, that makes sense.

Reply:
answers my question as well.